File: packages/core/loops/work-graph.ts
export type WorkGraphStatus =
  | 'planning'
  | 'running'
  | 'blocked'
  | 'awaiting_approval'
  | 'released'
  | 'observing'
  | 'completed'
  | 'failed'
  | 'cancelled'

export type WorkGraphNodeKind =
  | 'agent'
  | 'evaluation'
  | 'approval'
  | 'release'
  | 'observe'

export interface WorkGraph {
  id: string
  workspace_id: string
  project_id: string
  intent: WorkGraphIntent
  status: WorkGraphStatus
  current_stage?: string
  nodes: WorkGraphNode[]
  edges: WorkGraphEdge[]
  gates: WorkGraphGate[]
  provenance_summary: WorkGraphProvenanceSummary
}

export interface WorkGraphIntent {
  title: string
  description?: string
  source?: string
}

export interface WorkGraphNode {
  key: string
  kind: WorkGraphNodeKind
  role?: string
  responsibility?: string
  stage: string
  state: string
  assignee?: WorkGraphAssignee
  current_task?: WorkGraphTask
  evaluation?: WorkGraphEvaluation
  artifact_count: number
}

export interface WorkGraphAssignee {
  type: 'agent' | 'squad' | 'human'
  id: string
  name?: string
}

export interface WorkGraphTask {
  id: string
  status: string
  attempt?: number
}

export interface WorkGraphEvaluation {
  verdict: string
  score?: number
  finding_count: number
}

export interface WorkGraphEdge {
  from: string
  to: string
  kind: 'dependency' | 'recovery' | 'gate'
}

export interface WorkGraphGate {
  node_key: string
  type: 'human_approval' | 'evaluation'
  state: string
}

export interface WorkGraphProvenanceSummary {
  artifact_count: number
  evaluation_count: number
}

File: server/internal/loopservice/work_graph.go
package loopservice

import "context"

// WorkGraph is the product-facing projection of a Manifold Agent Native Loop.
// It intentionally hides Multica persistence/runtime details from API consumers.
type WorkGraph struct {
	ID                string                 `json:"id"`
	WorkspaceID       string                 `json:"workspace_id"`
	ProjectID         string                 `json:"project_id"`
	Intent            WorkGraphIntent        `json:"intent"`
	Status            string                 `json:"status"`
	CurrentStage      string                 `json:"current_stage,omitempty"`
	Nodes             []WorkGraphNode        `json:"nodes"`
	Edges             []WorkGraphEdge        `json:"edges"`
	Gates             []WorkGraphGate        `json:"gates"`
	ProvenanceSummary WorkGraphProvenance    `json:"provenance_summary"`
}

type WorkGraphIntent struct {
	Title       string `json:"title"`
	Description string `json:"description,omitempty"`
	Source      string `json:"source,omitempty"`
}

type WorkGraphNode struct {
	Key            string               `json:"key"`
	Kind           string               `json:"kind"`
	Role           string               `json:"role,omitempty"`
	Responsibility string               `json:"responsibility,omitempty"`
	Stage          string               `json:"stage"`
	State          string               `json:"state"`
	Assignee       *WorkGraphAssignee   `json:"assignee,omitempty"`
	CurrentTask    *WorkGraphTask       `json:"current_task,omitempty"`
	Evaluation     *WorkGraphEvaluation `json:"evaluation,omitempty"`
	ArtifactCount  int                  `json:"artifact_count"`
}

type WorkGraphAssignee struct {
	Type string `json:"type"`
	ID   string `json:"id"`
	Name string `json:"name,omitempty"`
}

type WorkGraphTask struct {
	ID      string `json:"id"`
	Status  string `json:"status"`
	Attempt int    `json:"attempt,omitempty"`
}

type WorkGraphEvaluation struct {
	Verdict      string   `json:"verdict"`
	Score        *float64 `json:"score,omitempty"`
	FindingCount int      `json:"finding_count"`
}

type WorkGraphEdge struct {
	From string `json:"from"`
	To   string `json:"to"`
	Kind string `json:"kind"`
}

type WorkGraphGate struct {
	NodeKey string `json:"node_key"`
	Type    string `json:"type"`
	State   string `json:"state"`
}

type WorkGraphProvenance struct {
	ArtifactCount   int `json:"artifact_count"`
	EvaluationCount int `json:"evaluation_count"`
}

// WorkGraphReader isolates the Manifold product contract from the concrete
// workforce/runtime implementation. Multica is the first projection backend.
type WorkGraphReader interface {
	GetWorkGraph(ctx context.Context, workspaceID, parentIssueID string) (*WorkGraph, error)
}
